add mergeLimit function as mentioned in #374

dayoadeyemi opened the pull request.

vqvu commented:
This breaks backpressure. You should be using modified version of the current implementation of `merge`. Then have `merge` return `this.mergeLimit(Infinity)`.

You all so need to add a `noValueOnErrorTest`. See the other tests for details.

vqvu commented on lib/index.js:
```diff
@@ -3611,6 +3611,78 @@ Stream.prototype.merge = function () {
+ *
+ * _([txt, md, js]).mergeLimit(2);
+ * // => contents of foo.txt, bar.txt, baz.txt and bosh.js in the order
+ * they were read, but bosh is not read until the others have comleted
```
Typo in "comleted". Also, this wording suggests bosh won't be read until both bar and baz have completed. I suggest "but bosh is not read until either foo/bar or baz has finished.

vqvu commented on lib/index.js:
```diff
@@ -3611,6 +3611,78 @@ Stream.prototype.merge = function () {
+ * @name Stream.mergeLimit()
+ * @api public
+ *
+ * var txt = _(['foo.txt', 'bar.txt']).map(readFile)
```
These all need to be `flatMap(readFile)`.

vqvu commented on lib/index.js:
```diff
@@ -3611,6 +3611,78 @@ Stream.prototype.merge = function () {
+ *
+ * @id mergeLimit
+ * @section Higher-order Streams
+ * @name Stream.mergeLimit()
```
Call this `Stream.mergeLimit(n)`. You're also missing a description of the parameter `n`.

vqvu commented:
Finally, let's call this `mergeWithLimit` to match the existing `batchWIthTimeOrCount`.

dayoadeyemi commented:
Yes that did break back-pressure, Instead of modifying the merge function I've made it so that there is just a transform function before the merge which prevents the stream from pushing any values until something from the current working set has ended.

vqvu commented on lib/index.js:
```diff
@@ -3611,6 +3611,78 @@ Stream.prototype.merge = function () {
+                push(err, x);
+                // console.log('start', x.id)
+                    x.once('end', function(){
+                    // console.log('end', x.id)
```
Your indentation for `x.once` is off. Also, I think you want to do `index--` in the `end` handler, right?

vqvu commented:
> I've made it so that there is just a transform function before the merge which prevents the stream from pushing any values until something from the current working set has ended.

Oh, that's a much better idea.

dayoadeyemi commented:
Yea if I don't index-- when the substream ends it will end up only allowing 1 process to run at a time. I've added that in and renamed index to processCount so that it is more descriptive.

vqvu commented on lib/index.js:
```diff
@@ -3611,6 +3611,79 @@ Stream.prototype.merge = function () {
+ *
+ * @id mergeLimit
+ * @section Higher-order Streams
+ * @name Stream.mergeWithLimit()
```
Call this `@name Stream.mergeWithLimit()`.

You also need to add a `@param n` field for the parameter.

vqvu commented on lib/index.js:
```diff
@@ -3611,6 +3611,79 @@ Stream.prototype.merge = function () {
+ *
+ * _([txt, md, js]).mergeWithLimit(2);
+ * // => contents of foo.txt, bar.txt, baz.txt and bosh.js in the order
+ * they were read, but bosh.js is not read until either foo.txt and bar.txt
```
This doesn't format correctly. Run `grunt docs` and open `docs/index.html` to see what I mean (don't commit it!).

Everything after `@api public` is treated as code, so you need a `//` before each of these lines.

vqvu commented on lib/index.js:
```diff
@@ -3611,6 +3611,79 @@ Stream.prototype.merge = function () {
+    if (n === Infinity) {
+        return this.merge();
+    } else if (n < 1) {
+        return _([]);
```
This should throw an error instead of returning an empty stream.

vqvu commented:
I think this is ready as soon as you fix a few issues with the documentation.

svozza commented:
A couple of stray `console.log` debug statements commented out but other than that this looks good to merge (pardon the pun).

vqvu commented:
Cool. Commented `console.log` are fine, since they're all over the rest of the codebase anyway.

Thanks, @dayoadeyemi for the PR!
